File: bindings/csharp/FusionFramework/Middleware.cs
using System.Text;
using System.Text.Json;
using System.Text.Json.Nodes;

namespace FusionFramework;

public delegate object? FusionMiddleware(FusionRequest request, Func<FusionRequest, object?> callNext);

public static class Middleware
{
    static readonly List<FusionMiddleware> ActiveGlobal = new();

    public static void SetActiveGlobal(IEnumerable<FusionMiddleware> middlewares)
    {
        ActiveGlobal.Clear();
        ActiveGlobal.AddRange(middlewares);
    }

    public static IReadOnlyList<FusionMiddleware> GetActiveGlobal() => ActiveGlobal.ToList();

    public static void ClearActiveGlobal() => ActiveGlobal.Clear();

    public static object? RunChain(
        FusionRequest request,
        IReadOnlyList<FusionMiddleware> middlewares,
        Func<FusionRequest, object?> handler)
    {
        EnsureState(request);

        object? Dispatch(int index, FusionRequest req)
        {
            if (index >= middlewares.Count)
                return handler(req);

            var mw = middlewares[index];
            return mw(req, next => Dispatch(index + 1, next));
        }

        // Sync FFI callback: unwrap Task/ValueTask from async handlers / middleware.
        return ResolveAwaitable(Dispatch(0, request));
    }

    /// <summary>
    /// Unwrap nested <see cref="Task"/> / <see cref="ValueTask"/> results (async API methods).
    /// Mirrors Python awaiting coroutines before treating the value as an HTTP body.
    /// </summary>
    public static object? ResolveAwaitable(object? result)
    {
        for (var depth = 0; depth < 8; depth++)
        {
            switch (result)
            {
                case null:
                    return null;
                case Task task:
                    {
                        task.ConfigureAwait(false).GetAwaiter().GetResult();
                        var t = task.GetType();
                        if (t.IsGenericType)
                        {
                            result = t.GetProperty("Result")?.GetValue(task);
                            continue;
                        }
                        return null;
                    }
                default:
                    {
                        var type = result.GetType();
                        if (type == typeof(ValueTask))
                        {
                            ((ValueTask)result).ConfigureAwait(false).GetAwaiter().GetResult();
                            return null;
                        }
                        if (type.IsGenericType && type.GetGenericTypeDefinition() == typeof(ValueTask<>))
                        {
                            // ValueTask<T> → Task<T> then unwrap on next iteration.
                            result = type.GetMethod("AsTask")!.Invoke(result, null);
                            continue;
                        }
                        return result;
                    }
            }
        }

        throw new InvalidOperationException("Awaitable nesting too deep");
    }

    public static Dictionary<string, JsonNode?> EnsureState(FusionRequest request)
    {
        return request.State;
    }

    /// <summary>Bearer JWT middleware. Optional <paramref name="verify"/>; otherwise base64url-decodes payload (no crypto).</summary>
    public static FusionMiddleware BearerJwt(
        Func<string, JsonObject?>? verify = null,
        string stateKey = "jwt",
        string header = "Authorization")
    {
        return (request, callNext) =>
        {
            request.Headers.TryGetValue(header, out var auth);
            if (string.IsNullOrEmpty(auth))
            {
                foreach (var kv in request.Headers)
                {
                    if (string.Equals(kv.Key, header, StringComparison.OrdinalIgnoreCase))
                    {
                        auth = kv.Value;
                        break;
                    }
                }
            }

            if (string.IsNullOrEmpty(auth) || !auth.StartsWith("Bearer ", StringComparison.OrdinalIgnoreCase))
                return Error(401, "Missing bearer token");

            var token = auth["Bearer ".Length..].Trim();
            try
            {
                JsonObject? payload;
                if (verify != null)
                {
                    payload = verify(token);
                    if (payload is null)
                        return Error(401, "Invalid token");
                }
                else
                {
                    var parts = token.Split('.');
                    if (parts.Length != 3) throw new InvalidOperationException("bad token");
                    var json = Encoding.UTF8.GetString(Base64UrlDecode(parts[1]));
                    payload = JsonNode.Parse(json) as JsonObject
                              ?? throw new InvalidOperationException("bad payload");
                }

                EnsureState(request)[stateKey] = payload;
                return callNext(request);
            }
            catch
            {
                return Error(401, "Invalid token");
            }
        };
    }

    public static FusionMiddleware RequireRoles(
        IEnumerable<string> roles,
        string claim = "roles",
        string stateKey = "jwt")
    {
        var allowed = new HashSet<string>(roles.Select(r => r.ToString()), StringComparer.Ordinal);
        return (request, callNext) =>
        {
            if (!EnsureState(request).TryGetValue(stateKey, out var payload) || payload is null)
                return Error(401, "Authentication required");

            var claimNode = payload[claim];
            if (claimNode is null)
                return Error(403, $"Missing '{claim}' claim");

            List<string> userRoles;
            if (claimNode is JsonArray arr)
                userRoles = arr.Select(n => n?.ToString() ?? "").ToList();
            else if (claimNode is JsonValue)
                userRoles = new List<string> { claimNode.ToString() };
            else
                return Error(403, $"Invalid '{claim}' claim");

            if (!userRoles.Any(r => allowed.Contains(r)))
            {
                return new Dictionary<string, object?>
                {
                    ["status"] = 403,
                    ["body"] = new Dictionary<string, object?>
                    {
                        ["detail"] = "Insufficient permissions",
                        ["required"] = allowed.ToList(),
                    },
                };
            }

            return callNext(request);
        };
    }

    public static FusionMiddleware RequireRoles(params string[] roles) =>
        RequireRoles((IEnumerable<string>)roles);

    /// <summary>Common security response headers.</summary>
    public static FusionMiddleware SecurityHeaders(
        string contentTypeOptions = "nosniff",
        string frameOptions = "DENY",
        string referrerPolicy = "strict-origin-when-cross-origin",
        string permissionsPolicy = "camera=(), microphone=(), geolocation=(), payment=()",
        string coop = "same-origin",
        string corp = "same-origin",
        string? csp = null,
        string? hsts = null)
    {
        var extra = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase)
        {
            ["X-Content-Type-Options"] = contentTypeOptions,
            ["X-Frame-Options"] = frameOptions,
            ["Referrer-Policy"] = referrerPolicy,
            ["Permissions-Policy"] = permissionsPolicy,
            ["Cross-Origin-Opener-Policy"] = coop,
            ["Cross-Origin-Resource-Policy"] = corp,
        };
        if (!string.IsNullOrEmpty(csp)) extra["Content-Security-Policy"] = csp!;
        if (!string.IsNullOrEmpty(hsts)) extra["Strict-Transport-Security"] = hsts!;

        return (request, callNext) =>
        {
            var result = ResolveAwaitable(callNext(request));
            return MergeResponseHeaders(result, extra);
        };
    }

    /// <summary>Set <c>Cache-Control</c> on responses.</summary>
    public static FusionMiddleware CacheHeaders(string defaultValue = "no-store") =>
        (request, callNext) =>
        {
            var result = ResolveAwaitable(callNext(request));
            return MergeResponseHeaders(result, new Dictionary<string, string>
            {
                ["Cache-Control"] = defaultValue,
            });
        };

    /// <summary>Echo or generate <c>X-Request-Id</c> on each request.</summary>
    public static FusionMiddleware RequestId(string header = "X-Request-Id", bool incoming = true) =>
        (request, callNext) =>
        {
            string? rid = incoming ? GetHeader(request, header) : null;
            if (string.IsNullOrEmpty(rid))
                rid = Guid.NewGuid().ToString();
            EnsureState(request)["request_id"] = rid;
            var result = ResolveAwaitable(callNext(request));
            return MergeResponseHeaders(result, new Dictionary<string, string> { [header] = rid });
        };

    /// <summary>CORS middleware; answers <c>OPTIONS</c> preflight with 204.</summary>
    public static FusionMiddleware Cors(
        IEnumerable<string>? allowOrigins = null,
        IEnumerable<string>? allowMethods = null,
        IEnumerable<string>? allowHeaders = null,
        IEnumerable<string>? exposeHeaders = null,
        bool allowCredentials = false,
        int maxAge = 600)
    {
        var origins = (allowOrigins ?? new[] { "*" }).Select(o => o.ToString()).ToList();
        var methods = (allowMethods ?? new[]
        {
            "GET", "POST", "PUT", "PATCH", "DELETE", "OPTIONS", "HEAD",
        }).Select(m => m.ToUpperInvariant()).ToList();
        var headers = (allowHeaders ?? new[]
        {
            "Authorization", "Content-Type", "Accept", "Origin", "X-Request-Id",
        }).ToList();
        var expose = (exposeHeaders ?? new[] { "X-Request-Id" }).ToList();
        var allowAll = origins.Contains("*");

        Dictionary<string, string> CorsHeaders(string? origin)
        {
            var chosen = "*";
            if (!allowAll)
            {
                if (!string.IsNullOrEmpty(origin) && origins.Contains(origin))
                    chosen = origin;
                else if (origins.Count > 0)
                    chosen = origins[0];
            }

            var map = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase)
            {
                ["Access-Control-Allow-Origin"] = chosen,
                ["Access-Control-Allow-Methods"] = string.Join(", ", methods),
                ["Access-Control-Allow-Headers"] = string.Join(", ", headers),
                ["Access-Control-Expose-Headers"] = string.Join(", ", expose),
                ["Access-Control-Max-Age"] = maxAge.ToString(),
                ["Vary"] = "Origin",
            };
            if (allowCredentials && chosen != "*")
                map["Access-Control-Allow-Credentials"] = "true";
            return map;
        }

        return (request, callNext) =>
        {
            var origin = GetHeader(request, "Origin");
            var extra = CorsHeaders(origin);
            if (string.Equals(request.Method, "OPTIONS", StringComparison.OrdinalIgnoreCase))
            {
                return new Dictionary<string, object?>
                {
                    ["status"] = 204,
                    ["body"] = "",
                    ["headers"] = extra,
                };
            }

            var result = ResolveAwaitable(callNext(request));
            return MergeResponseHeaders(result, extra);
        };
    }

    /// <summary>Optional identity middleware — not enabled by default. Add via <c>app.Use(Middleware.FrameworkHeaders())</c>.</summary>
    public static FusionMiddleware FrameworkHeaders()
    {
        var extra = Header.Fingerprint();
        return (request, callNext) =>
        {
            // Must resolve awaitables first — otherwise Task becomes the JSON body
            // (same class of bug as Python's un-awaited coroutine stringification).
            var result = ResolveAwaitable(callNext(request));
            return MergeResponseHeaders(result, extra);
        };
    }

    internal static object MergeResponseHeaders(object? result, IReadOnlyDictionary<string, string> extra)
    {
        if (result is Dictionary<string, object?> dict)
        {
            var headers = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
            foreach (var kv in extra) headers[kv.Key] = kv.Value;
            if (dict.TryGetValue("headers", out var existing) && existing is IDictionary<string, string> map)
            {
                foreach (var kv in map) headers[kv.Key] = kv.Value;
            }
            dict["headers"] = headers;
            return dict;
        }

        return new Dictionary<string, object?>
        {
            ["status"] = 200,
            ["body"] = result,
            ["headers"] = new Dictionary<string, string>(extra, StringComparer.OrdinalIgnoreCase),
        };
    }

    static object Error(int status, string detail) =>
        new Dictionary<string, object?>
        {
            ["status"] = status,
            ["body"] = new Dictionary<string, object?> { ["detail"] = detail },
        };

    static string? GetHeader(FusionRequest request, string name)
    {
        if (request.Headers.TryGetValue(name, out var direct) && !string.IsNullOrEmpty(direct))
            return direct;
        foreach (var kv in request.Headers)
        {
            if (string.Equals(kv.Key, name, StringComparison.OrdinalIgnoreCase))
                return kv.Value;
        }
        return null;
    }

    static byte[] Base64UrlDecode(string input)
    {
        var s = input.Replace('-', '+').Replace('_', '/');
        switch (s.Length % 4)
        {
            case 2: s += "=="; break;
            case 3: s += "="; break;
        }
        return Convert.FromBase64String(s);
    }
}

File: crates/fusion-py/python/fusion_framework/middleware.py
"""Global and route middleware for Fusion Framework."""

from __future__ import annotations

import base64
import inspect
import json
from pathlib import Path
from typing import Any, Awaitable, Callable, Iterable, Union

Middleware = Callable[..., Any]
RequestDict = dict[str, Any]
MaybeAwaitable = Union[Any, Awaitable[Any]]

_active_global: list[Middleware] = []


def use(middleware: Middleware) -> Middleware:
    """Register global middleware on the active app (called by ``FusionApp.use``)."""
    _active_global.append(middleware)
    return middleware


def set_active_global(middlewares: Iterable[Middleware]) -> None:
    global _active_global
    _active_global = list(middlewares)


def get_active_global() -> list[Middleware]:
    return list(_active_global)


def clear_active_global() -> None:
    _active_global.clear()


async def _maybe_await(value: MaybeAwaitable) -> Any:
    if inspect.isawaitable(value):
        return await value
    return value


def _ensure_state(request: RequestDict) -> dict[str, Any]:
    state = request.get("state")
    if not isinstance(state, dict):
        state = {}
        request["state"] = state
    return state


def _is_response(value: Any) -> bool:
    return isinstance(value, dict) and "status" in value


def _middleware_is_async(middleware: Middleware) -> bool:
    return inspect.iscoroutinefunction(middleware)


def _chain_needs_async(middlewares: list[Middleware], handler: Callable[..., Any]) -> bool:
    if inspect.iscoroutinefunction(handler):
        return True
    return any(_middleware_is_async(mw) for mw in middlewares)


def _run_chain_sync(
    request: RequestDict,
    middlewares: list[Middleware],
    handler: Callable[[RequestDict], Any],
) -> Any:
    _ensure_state(request)

    def dispatch(index: int, req: RequestDict) -> Any:
        if index >= len(middlewares):
            return handler(req)
        middleware = middlewares[index]

        def call_next(next_req: RequestDict) -> Any:
            return dispatch(index + 1, next_req)

        result = middleware(req, call_next)
        # Sync middleware may return a coroutine when the route handler is
        # ``async def`` (HandlerInvoker is sync but yields an awaitable).
        # Propagate to Rust so ``async_runtime.submit`` can await it.
        if inspect.isawaitable(result):
            return result
        if _is_response(result):
            return result
        return result

    return dispatch(0, request)


async def run_chain(
    request: RequestDict,
    middlewares: list[Middleware],
    handler: Callable[[RequestDict], MaybeAwaitable],
) -> Any:
    _ensure_state(request)

    async def dispatch(index: int, req: RequestDict) -> Any:
        if index >= len(middlewares):
            return await _maybe_await(handler(req))
        middleware = middlewares[index]

        async def call_next(next_req: RequestDict) -> Any:
            return await dispatch(index + 1, next_req)

        result = middleware(req, call_next)
        result = await _maybe_await(result)
        if _is_response(result):
            return result
        return result

    return await dispatch(0, request)


def require_permissions(*checks: Callable[[RequestDict], Any]) -> Middleware:
    """Route middleware: run custom permission checks; any falsy result → 403.

    Each check receives the request dict (``method``, ``path``, ``headers``,
    ``body``, ``state``, …). Omit checks to allow any caller (default).
    """

    def middleware(request: RequestDict, call_next: Callable[[RequestDict], Any]) -> Any:
        for check in checks:
            if not check(request):
                return {"status": 403, "body": {"detail": "Forbidden"}}
        return call_next(request)

    return middleware


def require_roles(
    *roles: str,
    claim: str = "roles",
    state_key: str = "jwt",
) -> Middleware:
    """Route middleware: allow only requests whose JWT/state payload includes a role."""
    allowed = {str(r) for r in roles}

    def middleware(request: RequestDict, call_next: Callable[[RequestDict], Any]) -> Any:
        state = _ensure_state(request)
        payload = state.get(state_key)
        if payload is None:
            return {"status": 401, "body": {"detail": "Authentication required"}}

        user_roles = payload.get(claim) if isinstance(payload, dict) else None
        if user_roles is None:
            return {"status": 403, "body": {"detail": f"Missing '{claim}' claim"}}

        if isinstance(user_roles, str):
            user_roles = [user_roles]
        if not isinstance(user_roles, (list, tuple, set)):
            return {"status": 403, "body": {"detail": f"Invalid '{claim}' claim"}}

        if not allowed.intersection({str(r) for r in user_roles}):
            return {
                "status": 403,
                "body": {
                    "detail": "Insufficient permissions",
                    "required": sorted(allowed),
                },
            }
        return call_next(request)

    return middleware


def bearer_jwt(
    *,
    state_key: str = "jwt",
    header: str = "Authorization",
    verify: Callable[[str], dict[str, Any] | None] | None = None,
) -> Middleware:
    """Extract Bearer token and store decoded payload in ``request['state']``."""

    def _decode_unverified(token: str) -> dict[str, Any] | None:
        try:
            parts = token.split(".")
            if len(parts) != 3:
                return None
            payload_b64 = parts[1] + "=" * (-len(parts[1]) % 4)
            raw = base64.urlsafe_b64decode(payload_b64.encode("ascii"))
            parsed = json.loads(raw.decode("utf-8"))
            return parsed if isinstance(parsed, dict) else None
        except (ValueError, json.JSONDecodeError, UnicodeDecodeError):
            return None

    header_keys = {header, header.lower(), header.upper()}

    def middleware(request: RequestDict, call_next: Callable[[RequestDict], Any]) -> Any:
        headers = request.get("headers") or {}
        auth = None
        for key, value in headers.items():
            if str(key) in header_keys:
                auth = value
                break

        if not auth or not str(auth).lower().startswith("bearer "):
            return {"status": 401, "body": {"detail": "Missing bearer token"}}

        token = str(auth)[7:].strip()
        payload = verify(token) if verify else _decode_unverified(token)
        if payload is None:
            return {"status": 401, "body": {"detail": "Invalid token"}}

        _ensure_state(request)[state_key] = payload
        return call_next(request)

    return middleware


def _merge_response_headers(result: Any, extra: dict[str, str]) -> Any:
    if not isinstance(result, dict):
        return {"status": 200, "body": result, "headers": dict(extra)}
    headers = result.get("headers")
    if not isinstance(headers, dict):
        headers = {}
    merged = {**extra, **headers}  # handler wins on conflict
    out = dict(result)
    out["headers"] = merged
    return out


def _call_next_merge_headers(
    call_next: Callable[[RequestDict], Any],
    request: RequestDict,
    extra: dict[str, str],
) -> Any:
    """Merge headers after ``call_next``, awaiting async handler results."""
    result = call_next(request)
    if inspect.isawaitable(result):

        async def _await_merge() -> Any:
            return _merge_response_headers(await result, extra)

        return _await_merge()
    return _merge_response_headers(result, extra)


def framework_headers() -> Middleware:
    """Optional identity middleware: advertise Fusion on responses.

    Adds ``X-Powered-By``, ``X-Framework``, and ``X-Fusion-Version``.
    Not enabled by default — add with ``app.use(framework_headers())``.
    Wire-level injection is off unless ``fingerprint.enabled: true`` in settings.
    """
    try:
        from fusion_framework._fusion import fingerprint_headers as _fp

        extra = dict(_fp())
    except Exception:
        try:
            from fusion_framework import header as hdr

            extra = {
                getattr(hdr, "X_POWERED_BY", "X-Powered-By"): getattr(
                    hdr, "FRAMEWORK_POWERED_BY", "Fusion Framework"
                ),
                getattr(hdr, "X_FRAMEWORK", "X-Framework"): getattr(hdr, "FRAMEWORK_ID", "Fusion"),
                getattr(hdr, "X_FUSION_VERSION", "X-Fusion-Version"): "1.2.6",
            }
        except Exception:
            extra = {
                "X-Powered-By": "Fusion Framework",
                "X-Framework": "Fusion",
                "X-Fusion-Version": "1.2.6",
            }

    def middleware(request: RequestDict, call_next: Callable[[RequestDict], Any]) -> Any:
        return _call_next_merge_headers(call_next, request, extra)

    return middleware


def _get_header(request: RequestDict, name: str) -> str | None:
    headers = request.get("headers") or {}
    target = name.lower()
    for key, value in headers.items():
        if str(key).lower() == target:
            return str(value)
    return None


def security_headers(
    *,
    content_type_options: str = "nosniff",
    frame_options: str = "DENY",
    referrer_policy: str = "strict-origin-when-cross-origin",
    permissions_policy: str = "camera=(), microphone=(), geolocation=(), payment=()",
    coop: str = "same-origin",
    corp: str = "same-origin",
    csp: str | None = None,
    hsts: str | None = None,
) -> Middleware:
    """Add common security response headers."""
    extra: dict[str, str] = {
        "X-Content-Type-Options": content_type_options,
        "X-Frame-Options": frame_options,
        "Referrer-Policy": referrer_policy,
        "Permissions-Policy": permissions_policy,
        "Cross-Origin-Opener-Policy": coop,
        "Cross-Origin-Resource-Policy": corp,
    }
    if csp:
        extra["Content-Security-Policy"] = csp
    if hsts:
        extra["Strict-Transport-Security"] = hsts

    def middleware(request: RequestDict, call_next: Callable[[RequestDict], Any]) -> Any:
        return _call_next_merge_headers(call_next, request, extra)

    return middleware


def cache_headers(*, default: str = "no-store") -> Middleware:
    """Set ``Cache-Control`` on responses that do not already define it."""
    extra = {"Cache-Control": default}

    def middleware(request: RequestDict, call_next: Callable[[RequestDict], Any]) -> Any:
        return _call_next_merge_headers(call_next, request, extra)

    return middleware


def request_id(
    *,
    header: str = "X-Request-Id",
    incoming: bool = True,
) -> Middleware:
    """Attach a request id to ``state`` and echo it on the response."""
    import uuid

    def middleware(request: RequestDict, call_next: Callable[[RequestDict], Any]) -> Any:
        state = _ensure_state(request)
        rid = _get_header(request, header) if incoming else None
        if not rid:
            rid = str(uuid.uuid4())
        state["request_id"] = rid
        return _call_next_merge_headers(call_next, request, {header: rid})

    return middleware


def cors(
    *,
    allow_origins: Iterable[str] | str = "*",
    allow_methods: Iterable[str] | None = None,
    allow_headers: Iterable[str] | None = None,
    expose_headers: Iterable[str] | None = None,
    allow_credentials: bool = False,
    max_age: int = 600,
) -> Middleware:
    """CORS middleware; short-circuits ``OPTIONS`` preflight with 204."""
    origins = (
        [str(allow_origins)]
        if isinstance(allow_origins, str)
        else [str(o) for o in allow_origins]
    )
    methods = (
        ["GET", "POST", "PUT", "PATCH", "DELETE", "OPTIONS", "HEAD"]
        if allow_methods is None
        else [str(m).upper() for m in allow_methods]
    )
    headers = (
        ["Authorization", "Content-Type", "Accept", "Origin", "X-Request-Id"]
        if allow_headers is None
        else [str(h) for h in allow_headers]
    )
    expose = ["X-Request-Id"] if expose_headers is None else [str(h) for h in expose_headers]
    allow_all = "*" in origins

    def _cors_headers(origin: str | None) -> dict[str, str]:
        chosen = "*"
        if not allow_all:
            if origin and origin in origins:
                chosen = origin
            elif origins:
                chosen = origins[0]
        out = {
            "Access-Control-Allow-Origin": chosen,
            "Access-Control-Allow-Methods": ", ".join(methods),
            "Access-Control-Allow-Headers": ", ".join(headers),
            "Access-Control-Expose-Headers": ", ".join(expose),
            "Access-Control-Max-Age": str(max_age),
            "Vary": "Origin",
        }
        if allow_credentials and chosen != "*":
            out["Access-Control-Allow-Credentials"] = "true"
        return out

    def middleware(request: RequestDict, call_next: Callable[[RequestDict], Any]) -> Any:
        origin = _get_header(request, "Origin")
        extra = _cors_headers(origin)
        if str(request.get("method", "GET")).upper() == "OPTIONS":
            return {"status": 204, "body": "", "headers": extra}
        return _call_next_merge_headers(call_next, request, extra)

    return middleware


_STATIC_MIME_TYPES: dict[str, str] = {
    ".css": "text/css; charset=utf-8",
    ".gif": "image/gif",
    ".htm": "text/html; charset=utf-8",
    ".html": "text/html; charset=utf-8",
    ".ico": "image/x-icon",
    ".jpeg": "image/jpeg",
    ".jpg": "image/jpeg",
    ".js": "text/javascript; charset=utf-8",
    ".json": "application/json",
    ".map": "application/json",
    ".png": "image/png",
    ".svg": "image/svg+xml",
    ".txt": "text/plain; charset=utf-8",
    ".webp": "image/webp",
    ".woff": "font/woff",
    ".woff2": "font/woff2",
}


def _guess_static_content_type(path: Path) -> str:
    """Map a file extension to a Content-Type, defaulting to octet-stream."""
    return _STATIC_MIME_TYPES.get(path.suffix.lower(), "application/octet-stream")


def static_files(
    root: str | Path = "static",
    *,
    prefix: str = "/static",
    max_age: int | None = 3600,
    fallthrough: bool | None = None,
) -> Middleware:
    """Serve local files under ``root`` for URLs starting with ``prefix`` (WhiteNoise-style).

    - ``root``: folder on disk that holds the files (e.g. ``\"static\"`` or
      ``Path(__file__).parent / \"templates\" / \"home\"``).
    - ``prefix``: URL path prefix browsers request (e.g. ``\"/static\"`` so
      ``static/logo.png`` is served at ``/static/logo.png``). Use ``\"/\"`` to
      serve files at the site root (``templates/home/a.png`` → ``/a.png``).

    On ``FusionApp.listen()``, matching files are mounted as real GET/HEAD routes
    (middleware alone cannot see unmatched paths). Only ``GET`` / ``HEAD``.
    Path traversal is rejected. The returned middleware also short-circuits when
    invoked on an already-mounted path.
    """
    root_path = Path(root).expanduser()
    normalized = "/" + str(prefix).strip("/") if str(prefix).strip("/") else "/"
    allow_fallthrough = (normalized == "/") if fallthrough is None else bool(fallthrough)
    cfg = {
        "root": root_path,
        "prefix": normalized,
        "max_age": max_age,
        "fallthrough": allow_fallthrough,
    }

    def middleware(request: RequestDict, call_next: Callable[[RequestDict], Any]) -> Any:
        return _serve_static_or_next(cfg, request, call_next)

    middleware.__fusion_static__ = cfg  # type: ignore[attr-defined]
    return middleware


def _serve_static_or_next(
    cfg: dict[str, Any],
    request: RequestDict,
    call_next: Callable[[RequestDict], Any],
) -> Any:
    """Try to serve a file for this request; otherwise call the next handler."""
    method = str(request.get("method", "GET")).upper()
    if method not in ("GET", "HEAD"):
        return call_next(request)

    req_path = str(request.get("path") or "/")
    normalized = str(cfg["prefix"])
    if normalized == "/":
        relative = req_path.lstrip("/")
        if not relative or relative.endswith("/"):
            return call_next(request)
    else:
        if not (req_path == normalized or req_path.startswith(normalized + "/")):
            return call_next(request)
        relative = req_path[len(normalized) :].lstrip("/")
        if not relative:
            return call_next(request)

    base = Path(cfg["root"]).resolve()
    candidate = (base / relative).resolve()
    try:
        candidate.relative_to(base)
    except ValueError:
        return {"status": 403, "body": {"detail": "Forbidden"}}

    if not candidate.is_file():
        if cfg["fallthrough"]:
            return call_next(request)
        return {"status": 404, "body": {"detail": "Not found"}}

    return _static_file_response(candidate, method, cfg.get("max_age"))


def _static_file_response(path: Path, method: str, max_age: int | None) -> dict[str, Any]:
    """Build a 200 response envelope for a file on disk."""
    size = path.stat().st_size
    headers = {
        "content-type": _guess_static_content_type(path),
        "content-length": str(size),
    }
    if max_age is not None:
        headers["cache-control"] = f"public, max-age={int(max_age)}"
    body: Any = b"" if method.upper() == "HEAD" else path.read_bytes()
    return {"status": 200, "body": body, "headers": headers}


def mount_static_files(engine: Any, middlewares: Iterable[Middleware]) -> None:
    """Register GET/HEAD routes for every file under each ``static_files`` mount."""
    for middleware in middlewares:
        cfg = getattr(middleware, "__fusion_static__", None)
        if not isinstance(cfg, dict):
            continue
        root = Path(cfg["root"]).expanduser().resolve()
        if not root.is_dir():
            continue
        prefix = str(cfg["prefix"])
        max_age = cfg.get("max_age")
        for file_path in root.rglob("*"):
            if not file_path.is_file():
                continue
            rel = file_path.relative_to(root).as_posix()
            url = f"/{rel}" if prefix == "/" else f"{prefix}/{rel}"
            path_for_get = file_path
            path_for_head = file_path

            def _get(_req: RequestDict, p: Path = path_for_get, age: int | None = max_age) -> dict[str, Any]:
                return _static_file_response(p, "GET", age)

            def _head(_req: RequestDict, p: Path = path_for_head, age: int | None = max_age) -> dict[str, Any]:
                return _static_file_response(p, "HEAD", age)

            engine.route("GET", url, _get)
            engine.route("HEAD", url, _head)


def dispatch_route(
    request: RequestDict,
    handler: Callable[[RequestDict], Any],
    route_middleware: list[Middleware],
) -> Any:
    """Entry from Rust: global middleware → route middleware → API handler."""
    chain = get_active_global() + list(route_middleware)

    if _chain_needs_async(chain, handler):
        return run_chain(request, chain, handler)
    return _run_chain_sync(request, chain, handler)
